donkeycar/parts/dgym.py

```diff
@@ -34,24 +35,48 @@ def __init__(self, sim_path, host="127.0.0.1", port=9091, headless=0, env_name="
                      'gyro': (0., 0., 0.),
                      'accel': (0., 0., 0.),
                      'vel': (0., 0., 0.)}
-        self.delay = float(delay)
+        self.delay = float(delay) / 1000
         self.record_location = record_location
         self.record_gyroaccel = record_gyroaccel
         self.record_velocity = record_velocity
         self.record_lidar = record_lidar
 
+        self.buffer = []
+
+    def delay_buffer(self, frame, info):
+        now = time.time()
+        buffer_tuple = (now, frame, info)
+        self.buffer.append(buffer_tuple)
+
+        # go through the buffer
+        to_remove = []
+        for it in range(len(self.buffer)):
+            if now-self.buffer[it][0] >= self.delay:
+                to_remove.append(it)
+                self.frame = self.buffer[it][1]
+            else:
+                break
+
+        # clear the buffer
+        for _ in range(len(to_remove)):
```

Comment:
Why do you need a loop here, it looks like you are deleting the same thing again and again.

Reply:
if there is multiple images "old enough" in the buffer, it will delete them from the buffer

Reply:
as they are in a list, removing [0] will remove the oldest

Reply:
You are right. This will pop `len(to_remove)` times from the front. I'm still a bit confused about what the `to_remove` array is meant to be doing? Aren't you just counting in line 55 and then popping `self.buffer` that count times from the front? So couldn't `to_remove` be just an `int` here that we count up in line 55 and then in 62 we just say `del self.buffer[:to_remove]` ?

Reply:
Yes you are right, I will refactor that